### DISCUSSION TITLE
How do I add colors to my lines?

### DISCUSSION THREAD

**Discussion by codingHippo** [tui-cs/Terminal.Gui#3391] [Q&A] [answered]
I am making an application that reads in files from a filesystem and displays them (this I have)
but based on certain rules I wish to highlight each line according the ruleset.

I am trying to figure out how I can have lines with different colors in the ListView or  TextView.
I have tried various properties wth Terminal.GUI without solving it, 
and I have various soltuions using ANSI codes more directly but each time I just get hte escape codes
printed instead of the color change. 

I want to be able to do something like this: 

(yellow) Lorem ipsum dolor (/yellow)
(blue)     sit amet, consectetur  (/blue)
(blue)    adipiscing elit, sed do  (/blue)
(blue)    eiusmod tempor incididunt  (/blue)
(green)  ut labore et dolore magna (/green)
(blue)    Ut enim ad minim  ((blue)
(pink)    veniam, quis nostrud (/pink)

The colors here are only meant to signify the color fo the line. 
I am not looking for markup syntax or anything like that.

**Comment by tznind** [comment]
TreeView has 'ColorGetter' and 'AspectGetter' that could achieve this.  If
you set the builder to return empty (i.e. no children) then it will
function like list view.

You can add objects to the tree and decide color using the above delegates.

Alternatively you can look at the syntax highlighting scenario or do
something custom drawing.

But there is no markup and trying to use ansii escapes directly will not go
well.  Terminal.Gui has a Driver layer which abstracts away color codes
into Attribute and ColorScheme so any attempt at drawing with codes
directly would mean fighting against the API instead of working with it.

**Reply by codingHippo** [reply]
Thank you. I will try this.

**Comment by tznind** [comment]
To work with list view custom rendering you would want to
implement IListDataSource and particularly the Render method.

To get started you can copy code from ListWrapper

https://github.com/gui-cs/Terminal.Gui/blob/develop/Terminal.Gui/Views/ListView.cs#L817

**Reply by codingHippo** [reply]
That is great thank you.

**Comment by BDisp** [comment, marked as answer]
The UICatalog has the "List View With Selection" scenario which uses different colors for even and odd lines. You can implement your own condition for display the wanted color. Take a look.

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/95ac8c0f-3713-4294-84cd-e29e9f4cc2d4)

**Reply by codingHippo** [reply]
Thank you for that.
